### DISCUSSION TITLE
Adding an Array of Attributes seem to be causing an issue as it throws an Error trying to WRITE a DCM FILE

### DISCUSSION THREAD

**Discussion by Homersbm** [dcm4che/dcm4che#924] [Q&A] [answered]
Adding an Array of Attributes seem to be causing an issue as it throws an Error trying to WRITE a DCM FILE
Please suggest what is the best way to add an Array of Attributes.

JSONReader readerStl = new JSONREADER( some InputStream)
 Attributes stlAttribute = new Attributes();
 readerStl.readDataset(stlAttribute);
 Sequence roiSeq; //

Attributes[] temp = new Attributes[2]; 
Attributes  ss = new Attributes();
ss.setValue(Tag.ContourSequence, VR.SQ,**temp**);  // **I want to ADD an Array of Attributes here but this line causes an Error. 
                                                                                            The Error is thrown at the time of WriteDataSet**
ss.setInt(Tag.ReferencedROINumber, VR.IS,1);
roiSeq.set(0,ss);
 
stlAttribute.setValue(Tag.ROIContourSequence, VR.SQ, roiSeq);

File dest = new File(contourFileName);
DicomOutputStream dos = new DicomOutputStream(dest);
dos.writeDataset(null, stlAttribute);

**Comment by gunterze** [comment, marked as answer]
`Attributes::setValue(tag, vr, value)` is for internal use only - there are more convenient methods for adding/overwriting attributes of a data set. Particularly for adding a Sequence attribute with items you have to use
```java
Sequence seq = attrs.newSequence(seg.ContourSequence, 2);
Attributes item1 = ...
Attributes item2 = ...
seq.add(item1);
seq.add(item2);
```

**Reply by Homersbm** [reply]
![image](https://user-images.githubusercontent.com/11217644/111917075-04849700-8aa4-11eb-82ef-487c429a90e2.png)

I have tried the approach that you have mentioned, however it says java.lang.UnsupportedOperationException.
Here is a Code snippet. If you could help point out where Exactly am doing this wrong, it would be great.
I have attached an Image of how  Tag.ContourSequence needs to be added to Tag.ROIContourSequence.

 Attributes temp = new Attributes();
      Sequence seq = temp.newSequence(Tag.ContourSequence, ContourSequenceList.size());
      ContourSequenceList.forEach( contour -> {
          seq.add(contour);
      });
      
  **stlAttribute_ContourDicomObject.setValue(Tag.ROIContourSequence, VR.SQ, seq);**
 File dest = new File(contourFileName);
    DicomOutputStream dos = new DicomOutputStream(dest);
    dos.writeDataset(null, stlAttribute_ContourDicomObject);  // Exception Thrown here
       
        
    
    ------*************---------------------
    How ContourSequenceList was formed earlier:   

           if (roiObj.numberOfROI > 0) {
                    String sopInstanceUid = att.getString(Tag.SOPInstanceUID);
                    String sopClassUid = att.getString(Tag.SOPClassUID);

                    for(int i = 0 ; i < roiObj.numberOfROI ; i++) {
                        Attributes coutourImageSeqItem = new Attributes();
                        coutourImageSeqItem.setString(Tag.ReferencedSOPInstanceUID,VR.UI,sopInstanceUid);
                        coutourImageSeqItem.setString(Tag.ReferencedSOPClassUID,VR.UI,sopClassUid);

                        Attributes[] contourImageSequence = new Attributes[1];
                        contourImageSequence[0] = coutourImageSeqItem;

                        Attributes contour = new Attributes();
                        contour.setValue(Tag.ContourImageSequence, VR.SQ, contourImageSequence);
                        contour.setValue(Tag.ContourData, VR.DS, roiObj.ptrToROIListContaingROI[i]);
                        contour.setInt(Tag.NumberOfContourPoints, VR.IS, roiObj.ptrToROIListContaingROI[i].length/3);
                        contour.setString(Tag.ContourGeometricType,VR.CS, "CLOSED_PLANAR");
                        ContourSequenceList.add(contour);
                    }
                   
                }

**Reply by gunterze** [reply]
You are stlll using
```java
                    Attributes[] contourImageSequence = ...;
                    contour.setValue(Tag.ContourImageSequence, VR.SQ, contourImageSequence);
```
You should also replace
```
contour.setValue(Tag.ContourData, VR.DS, roiObj.ptrToROIListContaingROI[i]);
```
by
```java
contour.set<Type>(...)
```

**Reply by Homersbm** [reply]
Thanks for the input. I am able to resolve my issue. I didnt grasp the basic  concept of Sequence earlier.
